### PR TITLE
Wells Improvements

### PR DESCRIPTION
**Pull request by dmclain**
The goal of this merge request is to improve the semantics of wells so that they are iterable, rendering degrades gracefully (ie you have choices between {{ well.render }} and {% for node in well.nodes %} ...lots of code... {% endfor %}

### CODE REVIEW

File: armstrong/core/arm_wells/models.py [hunk 1]
```diff
@@ -40,33 +48,137 @@ class Well(models.Model):
...
-                    content._meta.app_label,
-                    content._meta.object_name.lower(),
-                    self.type.slug), **kwargs))
-        return ''.join(ret)
-
+        if parent == self.well:
+            render_args['dictionary']['parent'] = None
+        if request:
+            render_args['context_instance'] = RequestContext(request)
+        return render_to_string("wells/%s/%s/%s.html" % (
```
Comment (niran) [inline armstrong/core/arm_wells/models.py]: This strikes me as something that will be inflexible in practice. Most Django code that uses automatically generated template names falls back to less specific names instead of requiring the most specific one outright. We should be using a list of templates like `["wells/<app>/<model>/<well>.html", "wells/<app>/<model>.html", "wells/<app>.html", "wells/_default.html"]`

Nothing specific to this pull request, but it's as convenient a place as any to point it out.

File: armstrong/core/arm_wells/models.py [hunk 2]
```diff
@@ -40,33 +48,137 @@ class Well(models.Model):
...
+                model_class = node.content_type.model_class()
+                if model_class == Well:
+                    managers, ordering = process_well(node.content_object,
+                            managers, ordering)
+                    continue
+                key = "%s.%s" % (model_class.__module__, node.content_type.model)
+                if not key in managers:
+                    managers[key] = {
+                            "name": node.content_type.model,
+                            "manager": model_class.objects,
```
Comment (niran) [inline armstrong/core/arm_wells/models.py]: IIRC, the default manager doesn't have to be called `objects`, and there's an attribute somewhere that always points to the default manager.

File: armstrong/core/arm_wells/models.py [hunk 3]
```diff
@@ -40,33 +48,137 @@ class Well(models.Model):
...
+                ordering[node_key] = i, well
+                managers[key]["object_ids"].append(node.object_id)
+                if self.queryset is not None and self.queryset.model is model_class:
+                    self.exclude_ids.append(node.object_id)
+            return managers, ordering
+
+        # well_managers {'module.model':name, manager, ids}
+        self.exclude_ids = []
+        well_managers, ordering = process_well(self, {}, {})
+        self.well_content = [i for i in range(len(ordering))]
```
Comment (niran) [inline armstrong/core/arm_wells/models.py]: This looks like you're trying to initialize a list with a length of ordering. If that's the case, it was a little difficult to determine. I think something like `[None] * len(ordering)` is a clearer way to do that.

**Comment by niran** [comment]
I'm a big fan of both of the goals of this pull request, and it looks well done to me.

File: armstrong/core/arm_wells/models.py (hunk 2 above)
Comment (dmclain) [inline armstrong/core/arm_wells/models.py]: You're correct `_default_manager` is the default manager always, though, as its name implies, it's an undocumented private type thing

File: armstrong/core/arm_wells/models.py (hunk 1 above)
Comment (dmclain) [inline armstrong/core/arm_wells/models.py]: The latest code does `wells/<app>/<model>/<well>.html` for each model in an object's class hierarchy and finally to `wells/<well>.html`. We could add `wells/_default.html` pretty easily too. The philosophy behind this is that if somebody had some special case class like FeaturedImage that subclasses a type that already has a template (in this case Image) it is better to use the super type's template than to error out or use some type of other default. 

Having a per-well default rather than a per-model default is for stylistic purposes. On TT's data page we have a well for popular data apps, if somebody threw an image in there accidentally, it could be bad for page layout if we just blindly rendered the default image layout, but it's not so bad if we use a well default that just renders the title.

File: armstrong/core/arm_wells/models.py [hunk 4]
```diff
@@ -40,33 +49,157 @@ class Well(models.Model):
...
+        if j <= total_in_well:
+            return [NodeWrapper(*args) for args in self.well_content[i:j]]
+        end = j - total_in_well
+        if i >= total_in_well:
+            start = i - total_in_well
+            return [NodeWrapper(content, self) for content in self.queryset[start:end]]
+        end = j - total_in_well
+        return itertools.chain(
+                (NodeWrapper(*args) for args in self.well_content[i:]),
+                (NodeWrapper(content, self) for content in self.queryset[0:end]))
```
Comment (tswicegood) [inline armstrong/core/arm_wells/models.py]: Does this need to be `0:end` or can it be `:end`?

File: armstrong/core/arm_wells/models.py [hunk 5]
```diff
@@ -40,33 +49,157 @@ class Well(models.Model):
...
+
+    def __getitem__(self, i):
+        if type(i) != type(1):
```
Comment (tswicegood) [inline armstrong/core/arm_wells/models.py]: This should really just check for `type(i) is not int`.  It's a cheaper comparison.

File: armstrong/core/arm_wells/models.py [hunk 6]
```diff
@@ -40,33 +49,157 @@ class Well(models.Model):
...
+    def count(self):
+        return self.__len__()
+
+    def __len__(self):
+        self.initialize()
+        return len(self.well_content) + \
+                self.queryset.count()
+
+    def __nonzero__(self):
+        return True
```
Comment (tswicegood) [inline armstrong/core/arm_wells/models.py]: I like the optimism, but is this _always_ True?

File: armstrong/core/arm_wells/models.py [hunk 7]
```diff
@@ -40,33 +49,157 @@ class Well(models.Model):
...
+
+    def __getattr__(self, key):
```
Comment (tswicegood) [inline armstrong/core/arm_wells/models.py]: Thoughts on whether we still need this method now that we're talking explicitly to an iterable model rather than a "queryset-like" object?

File: armstrong/core/arm_wells/models.py [hunk 8]
```diff
@@ -40,33 +49,157 @@ class Well(models.Model):
...
+        try:
+            return getattr(super(Well, self), key)
+        except AttributeError:
+            if key != 'queryset' and self.queryset is not None and hasattr(self.queryset, key):
+                raise NotImplementedError()
+            raise
+
+
+class NodeRenderMixin(object):
```
Comment (tswicegood) [inline armstrong/core/arm_wells/models.py]: This code feels out of place.  Why do we have rendering code in a mixin in `models`?

Comment (tswicegood) [inline armstrong/core/arm_wells/models.py]: And I sort of answered it myself by reading the rest of the code.  I wonder if there's a better name for it?

File: armstrong/core/arm_wells/models.py [hunk 9]
```diff
@@ -40,33 +49,157 @@ class Well(models.Model):
...
-                ret.append(render_to_string("wells/%s/%s/%s.html" % (
-                    content._meta.app_label,
-                    content._meta.object_name.lower(),
-                    self.type.slug), **kwargs))
-        return ''.join(ret)
-
-
-class Node(models.Model):
+                }
+        if parent == self.well:
```
Comment (tswicegood) [inline armstrong/core/arm_wells/models.py]: This logic is hard to follow here, so let me know if I've got this right.  We need to check for `parent` because we might be getting called inside a well that's inside another well and we want to check pass in "parent", but only when there's more than one well being used.  Correct?

File: armstrong/core/arm_wells/models.py [hunk 10]
```diff
@@ -77,5 +210,17 @@ class Node(models.Model):
...
+
+class NodeWrapper(NodeRenderMixin):
+    def __init__(self, content_object, well):
+        self.well = well
+        self.content_object = content_object
+
+    def __unicode__(self):
+        return u"%s: %s" % (self.well.title, self.content_object)
+
+    def __str__(self):
```
Comment (tswicegood) [inline armstrong/core/arm_wells/models.py]: Why `__str__` here but not in `Node`?

File: armstrong/core/arm_wells/tests/models.py [hunk 11]
```diff
@@ -66,13 +72,19 @@ class WellTestCase(TestCase):
                 story._meta.object_name.lower(), title)
         random_return = str(random.randint(1000, 2000))
 
-        render_to_string = fudge.Fake(callable=True)
-        context = {"well": well, "object": story, "parent": None}
-        render_to_string.with_args(expected_path, dictionary=context) \
-                .returns(random_return)
+        select_template = fudge.Fake(callable=True)
```
Comment (tswicegood) [inline armstrong/core/arm_wells/tests/models.py]: Just for future reference, there's been an API change in Fudge since this was originally written that deprecates `callable=True` in favor of `some_fake.is_callable()`.

File: armstrong/core/arm_wells/models.py (hunk 5 above)
Comment (niran) [inline armstrong/core/arm_wells/models.py]: I agree with your suggested change, but any compiler worth its salt will make that conversion at compile time.

File: armstrong/core/arm_wells/models.py (hunk 9 above)
Comment (dmclain) [inline armstrong/core/arm_wells/models.py]: I hate this logic so much, I just wrote it to match the unit tests as they were written originally, I assumed the parent semantics were complicated because you had a use case that needed it to be complicated.

Comment (niran) [inline armstrong/core/arm_wells/models.py]: It almost looks like the check is intended to avoid infinite well recursion, though it wouldn't actually achieve that. I can't see a reason why `parent` needs to be in the context.

Comment (niran) [inline armstrong/core/arm_wells/models.py]: Actually, I don't know what `parent` is at all.

File: armstrong/core/arm_wells/models.py [hunk 12]
```diff
@@ -40,33 +49,157 @@ class Well(models.Model):
...
+            parent = self
+        if request is not None:
+            context = RequestContext(request)
+        else:
+            context = Context()
+        kwargs = {'parent': parent,
+                  'context': context}
+        return mark_safe(''.join(n.render(**kwargs) for n in self))
+
+    def initialize(self):
```
Comment (tswicegood) [inline armstrong/core/arm_wells/models.py]: Per our conversation earlier, I think this needs to be refactored.  Let's pull the internal `process_well` code into a Manager's `flatten` method and move the de-dupe code into another method, probably on a `QuerySet`-like object.  All of the iteration code below should also be dispatched to that object so it's more independently testable.

File: armstrong/core/arm_wells/models.py (hunk 5 above)
Comment (tswicegood) [inline armstrong/core/arm_wells/models.py]: Compiler being the key word. :-)

File: armstrong/core/arm_wells/models.py (hunk 9 above)
Comment (tswicegood) [inline armstrong/core/arm_wells/models.py]: I think the original idea before moved here was that `parent` would be point to the well's parent.  That makes sense when the well is doing the rendering, but less so when the node is.  We could follow this with a `node` parameter in the context which would give us the ability to figure out what well was a part of it.  Might be worth refactoring.

File: armstrong/core/arm_wells/tests/query.py [hunk 13]
```diff
@@ -94,9 +94,9 @@ class MergedNodesAndQuerySetTest(TestCase):
 
         number_of_images, number_of_stories = 0, 0
         for node_model in node_models:
-            if node_model.__class__.__name__ == "Story":
+            if node_model.content_object.__class__.__name__ == "Story":
```
Comment (tswicegood) [inline armstrong/core/arm_wells/tests/query.py]: Aren't we changing the behavior of the iteration at this point?  The original code would return the actual related object.  Do we not want to do that now?

File: fabfile.py [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+    'TEMPLATE_CONTEXT_PROCESSORS': (
+        'django.core.context_processors.request',
+    ),
+    'ROOT_URLCONF': 'armstrong.core.arm_wells.tests.arm_wells_support.urls',
+}
+
+main_app = "arm_wells"
+full_name = "armstrong.core.arm_wells"
+tested_apps = (main_app, 'arm_wells_support')
+pip_install_first = True
```
Comment (tswicegood) [inline fabfile.py:19]: We can ditch this line now.

**Comment by tswicegood** [comment]
This is pretty close.  I'm going to wait until the refactor in `Well.initialize` happens though.

File: armstrong/core/arm_wells/models.py [hunk 15]
```diff
@@ -1,13 +1,13 @@
...
 from django.contrib.contenttypes.models import ContentType
-from django.template import RequestContext
-from django.template.loader import render_to_string
-from django.template import TemplateDoesNotExist
 
 from .managers import WellManager
-from .query import MergedNodesAndQuerySet
+from .querysets import MergeQuerySet, GenericForeignKeyQuerySet
+
+import itertools
```
Comment (tswicegood) [inline armstrong/core/arm_wells/models.py:10]: We can nix this.
